Readme enhancements (#11)
kcp-dev/kcp-tests: README.md

## Prerequisites
* Git installed. See [Installing Git](https://git-scm.com/book/en/v2/Getting-Started-Installing-Git)
* Golang installed. See [Installing Golang](https://golang.org/doc/install), the newer the better ( Ensure you install Golang from a binary release [found here](https://golang.org/dl/), not with a package manager such as `dnf` ）
* golint installed. See [Installing golint](https://github.com/golang/lint#installation)
* Install [kubelogin](https://github.com/int128/kubelogin.git) and [KCP kubectl plugin](https://github.com/kcp-dev/kcp.git) ( Ensure your installed KCP kubectl plugin version is the same with the kcp service version )

>   - If you would like to download the suitable version packages directly go to [kcp release assets](https://github.com/kcp-dev/kcp/releases), download the zip file suitable for your OS, extract and copy the binaries to a place which is in your PATH and once copied run `kubectl-kcp --version` to verify if server matches with client 

Reviewer: ```suggestion
  - If you would like to download the suitable version packages directly go to [kcp release assets](https://github.com/kcp-dev/kcp/tags), download the zip file suitable for your OS, extract and copy the binaries to a place which is in your PATH and once copied run `kubectl kcp --version` to verify if your client matches with the server 
```
It is better to use https://github.com/kcp-dev/kcp/tags. You see it can show a more latest one (currently v0.8.x) while https://github.com/kcp-dev/kcp/releases may show an outdated one (currently v0.7.x). Since this README will be for newcomers, if they follow https://github.com/kcp-dev/kcp/releases and use, they will possibly use an outdated one and then hit issue which is raised in Slack a few times which this suggestion can avoid.

Author: Hi @xingxingxia , the https://github.com/kcp-dev/kcp/releases also shows the latest one v0.8.x. Maybe `tags` not better for download as only release versions will build all archs plugins. E.g. https://github.com/kcp-dev/kcp/releases/tag/pkg%2Fapis%2Fv0.7.5 has no plugins assert as it's not a release version but it's in tags list. WDYT?
cc to @kasturinarra

Reviewer: OK, I did not scroll down in the page, its top shows v0.7.x ... so I thought ... : )

Author: Got it. Thanks!

  - If you would like to build the plugin by your self you could follow the steps like below:
  ```
  $ git clone https://github.com/kcp-dev/kcp.git

>   # E.g. Install the release-0.5 version KCP kubectl plugin
>   # commitID: "fccfb61" of release-0.5 branch/tag
>   $ git checkout fccfb61

Reviewer: ```suggestion
  # Check the latest version in https://github.com/kcp-dev/kcp/tags and figure out its commit ID to checkout.
  $ git checkout <commit ID>
```

Author: Addressed.

  $ make install WHAT=./cmd/kubectl-kcp && make install WHAT=./cmd/kubectl-workspaces && make install WHAT=./cmd/kubectl-ws
  ```

* Have the environment variable `KUBECONFIG` set pointing to your kcp service

> * Login the kcp service via SSO (Single Sign On)

Reviewer: ```suggestion
* Log in to the kcp service via SSO (Single Sign On)
```

Author: Addressed.

  ```

>   kubectl oidc-login get-token --oidc-issuer-url=https://sso.xxxx.com/auth
>    --oidc-client-id=cloud-services --oidc-redirect-url-hostname=127.0.0.1

Reviewer: AFAIK even the client ID is not proper to be public. And the URL example is not good enough. How about just generic:
```suggestion
  kubectl oidc-login get-token --oidc-issuer-url=<oidc issuer url>
   --oidc-client-id=<oidc client ID> --oidc-redirect-url-hostname=127.0.0.1
```

Author: Good point, addressed!

  ```

> * If you would like to test BYO cases please make sure to set the `PCLUSTER_KUBECONFIG` environment variable else all the tests related to BYO will be skipped. 

Reviewer: ```suggestion
* If you would like to test BYO cases please make sure the `PCLUSTER_KUBECONFIG` environment variable is set otherwise all the tests related to BYO will be skipped. 
```

Author: Addressed.

## Contribution 

> Below are the general steps for submitting a PR to master branch. First, you should **Fork** this repo to your own Github account.

Reviewer: ```suggestion
Below are the general steps for submitting a PR to main branch. First, you should **Fork** this repo to your own Github account.
```

Author: Addressed.

> ```console

Reviewer: Though seems to not have impact, why console is used? I see many occurrences.

Author: Hmm, it's origin fomat I don't change it, I think we could safely remove the `console`.

Reviewer: Ah, sorry, now I figured it out. No need to remove. With it, the non-command words will appear with different color.

Author: Good to know.

Reviewer: Sorry, with my last comment above, I mean keeping `console` can make non-command lines appear with good blue color like:
```console
$ make build
mkdir -p "bin"
export GO111MODULE="on" && export GOFLAGS="" && go build  -ldflags="-s -w" -mod=mod -o "bin" "./cmd/kcp-tests"
```
Now I see your https://github.com/Phaow/kcp-tests/blob/readme-enhancements/README.md kept other places of `console` and removed some places of `console`, e.g.:
```
$ make build
mkdir -p "bin"
export GO111MODULE="on" && export GOFLAGS="" && go build  -ldflags="-s -w" -mod=mod -o "bin" "./cmd/kcp-tests"
```
This may confuse newcomers to execute both mkdir and export. Could you add console back for those lines? Sorry if caused trouble.

Author: Hmm, I just replace `console` to `shell` for CLI execute items, only the log output keep `console`.
Use shell
![image](https://user-images.githubusercontent.com/86950592/191170330-311f383c-b191-4095-bf60-52e728575e6e.png)

Use console
![image](https://user-images.githubusercontent.com/86950592/191170442-86889a9e-cac5-40fb-9fc2-4c3a2e3cafaf.png)

Seems shell looks better for CLI execute. WDYT? OR we change all to `console`?

Author: Seems my local `md` plugin is not the same display with github. I changed all to `console` and remove the comment `#` for enhancing seems looks better now.

$ git remote add <Your Name> git@github.com:<Your Github Account>/kcp-tests.git
$ git pull origin main
$ git checkout -b <Branch Name>
$ git add xxx
$ git diff main --name-only |grep ".go$"| grep -v "bindata.go$" | xargs -n1 golint
  Please fix all golint error
$ git diff main --name-only |grep ".go$"| grep -v "bindata.go$" | xargs gofmt -s -l
  Please fix all gofmt error, running 'gofmt -s -d [file_path]' or autocorrect with 'gofmt -s -w [file_path]'
$ git add xxx
$ make build
$ ./bin/kcp-tests run all --dry-run |grep <Test Case Name>|./bin/kcp-tests run -f -
$ git commit -m "xxx"
$ git push <Your Name> <Branch Name>:<Branch Name>
```
And then there will be a prompt in your Github repo console to open a PR, click it to do so.
### Include new test folder

> If you create a new folder for your test case, **add the path** to the [include.go](https://github.com/openshift/openshift-tests-private/blob/master/test/extended/include.go)

Reviewer: ```suggestion
If you create a new folder for your test case, **add the path** to the [include.go](https://github.com/kcp-dev/kcp-tests/blob/main/test/extended/include.go)
```

Author: Good catch! Fixed.

### Create go-bindata for new YAML files
If you have some **new YAML files** used in your code, you have to generate the bindata first.
Run `make update` to update the bindata. For example, you can see the bindata has been updated after running the `make update` as follows:
```console
$ git status
	modified:   test/extended/testdata/bindata.go
	new file:   test/extended/testdata/kcp/xxxx.yaml
```

### Compile the executable binary
Note that we use the `go module` for package management, the previous `go path` is deprecated.
```console
$ git clone git@github.com:kcp-dev/kcp-tests.git
$ cd kcp-tests/
$ make build
mkdir -p "bin"
export GO111MODULE="on" && export GOFLAGS="" && go build  -ldflags="-s -w" -mod=mod -o "bin" "./cmd/kcp-tests"
$ ls -hl ./bin/kcp-tests 
-rwxrwxr-x. 1 cloud-user cloud-user 106M Sep 13 14:41 ./bin/kcp-tests
```

### Run the automation test case
The binary finds the test case via searching for the test case title. It searches the test case titles by RE (`Regular Expression`). So, you can filter your test cases by using `grep`. 

> ##### Run automation test cases related an area

Reviewer: ```suggestion
##### Run automation test cases related to an area
```

Author: Addressed.

If you want to run all smoke test cases which has the `[Smoke]` label, you can do:
```console
$ ./bin/kcp-tests run smoke
```
##### Run a single automation test case
If you want to run a single test case, such as `g.It("Author:pewang-Medium-[Smoke] Multi levels workspaces lifecycle should work"`, you can do:
```console
$ ./bin/kcp-tests run all --dry-run|grep "Multi levels workspaces lifecycle should work"|./bin/kcp-tests run --junit-dir=./ -f -
```

### Debugging
#### Keep generated temporary workspaces
Sometime, we want to **keep the generated workspaces for debugging**, we could just set **`export DELETE_WORKSPACE=false`**, then these temporary workspaces will be kept. 

```console
...
Dec 18 09:39:33.448: INFO: Running AfterSuite actions on all nodes
...
1 pass, 0 skip (1m50s)
$ kubectl get ws
NAME                        TYPE        PHASE   URL
e2e-test-kcp-syncer-s6ktv   universal   Ready   https://<kcp-test-env-domain>/clusters/root:users:rp:pv:rh-sso-xxxx:e2e-test-kcp-syncer-s6ktv
e2e-test-kcp-syncer-a5spq   universal   Ready   https://<kcp-test-env-domain>/clusters/root:users:rp:pv:rh-sso-xxxx:e2e-test-kcp-syncer-a5spq
...
```
#### Print cluster event on Terminal

> When you execute cases, there are some events which is printed to the terminal (**`currently we cannot retreive events from kcp`)**, like

Reviewer: Given we currently can't, how about adding TODO after 108 and commenting out all lines after 108? And if in future we can, we can remove the comments.

Author: Sure, added.

> <!-- TODO: Retreive events from kcp server by test framework-->

Reviewer: Seems this does not comment out line 108 ~ 124 from the preview of your https://github.com/Phaow/kcp-tests/blob/readme-enhancements/README.md .
Seems to need to use `//` from example of the raw https://raw.githubusercontent.com/openshift/openshift-docs/main/authentication/bound-service-account-tokens.adoc VS the preview https://github.com/openshift/openshift-docs/blob/main/authentication/bound-service-account-tokens.adoc

Author: Sorry, I misunstand that only add `TODO`. Addressed.

Author: Hmm, for `md` files seems use `<!-- -->` for comments lines is better.

Reviewer: Any one is fine as long as can make those lines hidden : )

* Golang installed. See [Installing Golang](https://golang.org/doc/install),the newer the better.
        * Ensure you install Golang from a binary release [found here](https://golang.org/dl/), not with a package manager such as `dnf`
* golint installed. See [Installing golint](https://github.com/golang/lint#installation)
* Install [kubelogin](https://github.com/int128/kubelogin.git) and [KCP kubectl plugin](https://github.com/kcp-dev/kcp.git)
      * Ensure your installed KCP kubectl plugin version is the same with the kcp service version, you could build the plugin by your self or just download the suitable version binary from [kcp release assets](https://github.com/kcp-dev/kcp/releases)

>   ```

Reviewer 2: @Phaow also may be two different sections, if you would like to build please follow the below procedure. If you woul dlike to download the suitable version binary go to this page, download the zip file suitable for your OS, extract and copy the binaries to a place which is in your PATH and once copied run /usr/bin/kubectl-kcp --version to verify if server matches with client ? Can we include this information ?

Author: Sure, addressed.

### Create go-bindata for new YAML files
If you have some **new YAML files** used in your code, you have to generate the bindata first.
Run `make update` to update the bindata. For example, you can see the bindata has been updated after running the `make update` as follows:
```console
$ git status
	modified:   test/extended/testdata/bindata.go
	new file:   test/extended/testdata/kcp/xxxx.yaml
```

### Compile the executable binary
Note that we use the `go module` for package management, the previous `go path` is deprecated.
```console
$ git clone git@github.com:kcp-dev/kcp-tests.git
$ cd kcp-tests/
$ make build
mkdir -p "bin"
export GO111MODULE="on" && export GOFLAGS="" && go build  -ldflags="-s -w" -mod=mod -o "bin" "./cmd/kcp-tests"
$ ls -hl ./bin/kcp-tests 
-rwxrwxr-x. 1 cloud-user cloud-user 106M Sep 13 14:41 ./bin/kcp-tests
```

### Run the automation test case
The binary finds the test case via searching for the test case title. It searches the test case titles by RE (`Regular Expression`). So, you can filter your test cases by using `grep`. 

> ##### Run an area automation test cases

Reviewer 2: @Phaow may be change this to Run automation cases related to an area ?

Author: Addressed.

* Golang installed. See [Installing Golang](https://golang.org/doc/install),the newer the better.
        * Ensure you install Golang from a binary release [found here](https://golang.org/dl/), not with a package manager such as `dnf`
* golint installed. See [Installing golint](https://github.com/golang/lint#installation)

> * Install [kubelogin](https://github.com/int128/kubelogin.git) and [KCP kubectl plugin](https://github.com/kcp-dev/kcp.git)

Reviewer 2: @Phaow may be it is better to ask them to install using the plugins ? that will be easy for everyone i believe ?

Author: Good option. I'll also add the simple way that download the package directly. In addition, I plan to write the dockfile and build the images with the release versions it will be more easily to run the test.

Author: Addressed.

Reviewer 2: @Phaow since the link for the plugin is given below, can we remove this link here otherwise it might be confusing for some one who is very new may be ?

      * Ensure your installed KCP kubectl plugin version is the same with the kcp service version
  ```
  $ git clone https://github.com/kcp-dev/kcp.git
  # E.g. Install the release-0.5 version KCP kubectl plugin
  # commitID: "fccfb61" of release-0.5 branch/tag
  $ git checkout fccfb61
  $ make install WHAT=./cmd/kubectl-kcp && make install WHAT=./cmd/kubectl-workspaces && make install WHAT=./cmd/kubectl-ws
  ```

* Have the environment variable `KUBECONFIG` set pointing to your kcp service
* Login the kcp service via SSO (Single Sign On)
  ```
  kubectl oidc-login get-token --oidc-issuer-url=https://sso.xxxx.com/auth
   --oidc-client-id=cloud-services --oidc-redirect-url-hostname=127.0.0.1
  ```

> * Have the environment variable `PCLUSTER_KUBECONFIG` set pointing to your BYO cluster(Bring your own cluster), it's optional, you should set it if your test cases need to operate BYO cluster. The test cases which need to operate BYO cluster will be skipped by no environment variable `PCLUSTER_KUBECONFIG` found without setting it.

Reviewer 2: @Phaow can we have this modified something like "If you would like to test BYO cases please make sure to set the PCLUSTER_KUBECONFIG else all the tests related to BYO will be skipped" ?

Author: Sure. I'll update it.

Author: Addressed.

```console
$ git status
	modified:   test/extended/testdata/bindata.go

> 	new file:   test/extended/testdata/kcp/xxxx.yaml

Reviewer 2: @Phaow this should come under contribution too.

Author: Addressed.

```

### Compile the executable binary
Note that we use the `go module` for package management, the previous `go path` is deprecated.
```console
$ git clone git@github.com:kcp-dev/kcp-tests.git
$ cd kcp-tests/
$ make build
mkdir -p "bin"
export GO111MODULE="on" && export GOFLAGS="" && go build  -ldflags="-s -w" -mod=mod -o "bin" "./cmd/kcp-tests"
$ ls -hl ./bin/kcp-tests 

> -rwxrwxr-x. 1 cloud-user cloud-user 106M Sep 13 14:41 ./bin/kcp-tests

Reviewer 2: This section should be after contribution

Author: Addressed.

If you want to run all smoke test cases which has the `[Smoke]` label, you can do:
```console
$ ./bin/kcp-tests run smoke
```
##### Run a single automation test case
If you want to run a single test case, such as `g.It("Author:pewang-Medium-[Smoke] Multi levels workspaces lifecycle should work"`, you can do:
```console
$ ./bin/kcp-tests run all --dry-run|grep "Multi levels workspaces lifecycle should work"|./bin/kcp-tests run --junit-dir=./ -f -
```

### Debugging
#### Keep generated temporary workspaces
Sometime, we want to **keep the generated workspaces for debugging**. Just add the Env Var: `export DELETE_WORKSPACE=false`. These random workspaces will be kept, like below:
```console
...
Dec 18 09:39:33.448: INFO: Running AfterSuite actions on all nodes
...
1 pass, 0 skip (1m50s)
$ kubectl get ws
NAME                        TYPE        PHASE   URL
e2e-test-kcp-syncer-s6ktv   universal   Ready   https://<kcp-test-env-domain>/clusters/root:users:rp:pv:rh-sso-xxxx:e2e-test-kcp-syncer-s6ktv
e2e-test-kcp-syncer-a5spq   universal   Ready   https://<kcp-test-env-domain>/clusters/root:users:rp:pv:rh-sso-xxxx:e2e-test-kcp-syncer-a5spq
...
```
#### Print cluster event on Terminal

> When you execute cases, there are some cluster event which is printed to the terminal(**`currently kcp not enable the events info`**), like

Reviewer 2: s/ currently kcp not enable the events info/ currently we cannot retreive events from kcp

Author: Addressed.

> In default, it does not print the cluster event on the terminal when you execute the case on your terminal.

Reviewer 2: @Phaow you should say "By default" ?

Author: Sure, I'll update it.

Author: Addressed.

> if you want to enable it for helping debug, **please set `export ENABLE_PRINT_EVENT_STDOUT=true` before executing the case.**

Reviewer 2: S/ if you want to enable it for helping debug/ If you would like to enable debugging for a case

Author: Addressed.
